app/controllers/guide_article_drafts_controller.ts

```diff
@@ -0,0 +1,286 @@
+    if (instance === null) {
+      throw new ForbiddenException();
+    }
+    const has = await user?.hasPermission?.(slug, instance);
+    if (has !== true) {
+      throw new ForbiddenException();
+    }
+  }
+
+  /**
+   * When linking to an existing article, require per-model permission on the original article.
+   *
+   * Note: Multiple drafts can be created for the same article. This is intentional to allow:
+   * - Different users to propose different changes to the same article
+   * - A user to create multiple draft versions before deciding which to submit
+   * - Parallel editing workflows
+   *
+   * The solvro_admin can approve/reject drafts individually, choosing which changes to apply.
+   */
+  protected async storeHook(ctx: {
+    http: HttpContext;
+    request: PartialModel<typeof GuideArticleDraft>;
+  }) {
+    const { http, request } = ctx;
+    const originalId = request.originalArticleId;
+    if (originalId === null || originalId === undefined) {
+      return;
+    }
+
+    const user = http.auth.user as unknown as
+      | {
+          hasRole?: (slug: string) => Promise<boolean>;
+          hasPermission?: (
+            action: string,
+            target?: unknown,
+          ) => Promise<boolean>;
+        }
+      | undefined;
+    const isAdmin =
+      (await user?.hasRole?.("solvro_admin")) === true ||
+      (await user?.hasRole?.("admin")) === true;
+    if (isAdmin) {
+      return;
+    }
+
+    const article = await GuideArticle.find(originalId);
+    if (article === null) {
+      throw new NotFoundException(
+        `GuideArticle with id ${originalId} not found`,
+      );
+    }
+    const allowed = await user?.hasPermission?.("update", article);
+    if (allowed !== true) {
+      throw new ForbiddenException();
+    }
+  }
+
+  $configureRoutes(
+    controller: LazyImport<
+      Constructor<BaseController<typeof GuideArticleDraft>>
+    >,
+  ) {
+    super.$configureRoutes(controller);
+    router
+      .post("/:id/approve", [
+        controller as LazyImport<Constructor<GuideArticleDraftsController>>,
+        "approve",
+      ])
+      .as("approve");
+  }
+
+  async approve({ request, auth }: HttpContext) {
+    if (!auth.isAuthenticated) {
+      await auth.authenticate();
+    }
+    // Only solvro_admin can approve drafts
+    const isSolvroAdmin = await (
+      auth.user as unknown as { hasRole?: (slug: string) => Promise<boolean> }
+    ).hasRole?.("solvro_admin");
+    if (isSolvroAdmin !== true) {
+      throw new ForbiddenException();
+    }
+
+    // Use autogenerated validator for ID
+    const {
+      params: { id: draftId },
+    } = (await request.validateUsing(this.pathIdValidator)) as {
+      params: { id: string | number };
+    };
+
+    // Use findOrFail with error context
+    const draft = await GuideArticleDraft.findOrFail(draftId).addErrorContext(
+      () =>
+        `GuideArticleDraft with id ${draftId} not found or user lacks permission`,
+    );
```

Comment:
the "user lacks permission" part seems just wrong
i'd get it if you would throw the exact same error on permission check failures, but you don't...